phc/easy/query/fhir_dsl_query.py

```diff
@@ -0,0 +1,87 @@
+from toolz import pipe, identity, curry
+from typing import List, Union
+from lenses import lens
+
+FHIR_WHERE = lens.Get("where", {})
+FHIR_WHERE_TYPE = FHIR_WHERE.Get("type", "")
+FHIR_SIMPLE_QUERY = FHIR_WHERE.Get("query", {})
+FHIR_BOOL_QUERY = FHIR_SIMPLE_QUERY.Get("bool", {})
+
+
+@curry
+def and_query_clause_terms(second_query_clause, first_query_clause):
+    return {
+        "bool": {
+            "should": [first_query_clause, second_query_clause],
+            "minimum_should_match": 2,
+        }
+    }
+
+
+def and_query_clause(query: dict, query_clause: dict):
+    "Append a term/terms clause to an existing FSS query"
+    if FHIR_WHERE.get()(query) == {}:
+        return pipe(
+            query,
+            FHIR_SIMPLE_QUERY.set(query_clause),
+            FHIR_WHERE_TYPE.set("elasticsearch"),
+        )
+
+    if FHIR_WHERE_TYPE.get()(query) != "elasticsearch":
+        raise ValueError(
+            "Could not add clause to query that is not elasticsearch",
+            query_clause,
+            query,
+        )
+
+    query_keys = list(FHIR_SIMPLE_QUERY.get()(query).keys())
+    bool_keys = FHIR_BOOL_QUERY.get()(query).keys()
+    if (len(query_keys) == 1 and (query_keys[0] in ["term", "terms"])) or (
+        "should" in bool_keys
+    ):
+        return FHIR_SIMPLE_QUERY.modify(and_query_clause_terms(query_clause))(
+            query
+        )
+
+    raise ValueError("Could not add clause to query", query_clause, query)
+
+
+def _patient_ids_adder(
+    patient_id: Union[str, None] = None,
+    patient_ids: List[str] = [],
```

Comment:
Avoid empty lists as default arguments. Python will create the default list when the function is loaded and keep reusing it. I would test this by invoking this function a few times, only providing `patient_id` and making sure `patient_ids` only contains that single value on each call.

Reply:
Thanks for the heads up. Based on https://docs.python-guide.org/writing/gotchas/, this is only going to be a problem if we're mutating the list since it will stay around and be associated with that function. One line 54, I re-assign the variable but don't mutate the list itself. I've also tested this in the way you mentioned, and it did not have problems.

```python
def test_no_weird_results():
    assert _patient_ids_adder("a")({})["where"]["query"]["terms"]["subject.reference.keyword"] == [
        "Patient/a", "a"
    ]

    assert _patient_ids_adder("b")({})["where"]["query"]["terms"]["subject.reference.keyword"] == [
        "Patient/b", "b"
    ]

    assert _patient_ids_adder("c")({})["where"]["query"]["terms"]["subject.reference.keyword"] == [
        "Patient/c", "c"
    ]

    assert _patient_ids_adder("c")({})["where"]["query"]["terms"]["subject.reference.keyword"] == [
        "Patient/c", "c"
    ]
```